Building Neovim fails on M1 Mac in both Rosetta and Native

I've obviously done something very wrong as I get similar error messages when I try to build manually in a native iTerm or install via homebrew -s in a rosetta iterm.

My system is a 16G M1 Mac Mini running iOS 11.0.1.

To build on apple silicon I did:
```
git clone git@github.com:neovim/neovim
cd neovim
make distclean
echo "DEPS_CMAKE_FLAGS += -DCMAKE_OSX_DEPLOYMENT_TARGET=11.00" > local.mk
make SDKROOT=`xcrun --show-sdk-path`
```

This gives error

```
ninja  -C build
ninja: Entering directory `build'
[234/291] Building C object src/nvim/CMakeFiles/nvim.dir/os/lang.c.o
FAILED: src/nvim/CMakeFiles/nvim.dir/os/lang.c.o
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -DINCLUDE_GENERATED_DECLARATIONS -DMIN_LOG_LEVEL=1 -DNVIM_MSGPACK_HAS_FLOAT32 -D_GNU_SOURCE -Dnvim_EXPORTS -Iconfig -I../src -Isrc/nvim/auto -Iinclude -I../.deps/usr/include/luajit-2.1 -isystem ../.deps/usr/include -isystem /usr/local/opt/gettext/include -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -g -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -MD -MT src/nvim/CMakeFiles/nvim.dir/os/lang.c.o -MF src/nvim/CMakeFiles/nvim.dir/os/lang.c.o.d -o src/nvim/CMakeFiles/nvim.dir/os/lang.c.o -c ../src/nvim/os/lang.c
../src/nvim/os/lang.c:63:5: error: implicit declaration of function 'setlocale' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    setlocale(LC_ALL, "");
    ^
../src/nvim/os/lang.c:63:15: error: use of undeclared identifier 'LC_ALL'; did you mean 'P_ALL'?
    setlocale(LC_ALL, "");
              ^~~~~~
              P_ALL
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/wait.h:80:2: note: 'P_ALL' declared here
        P_ALL,
        ^
2 errors generated.
[243/291] Building C object src/nvim/CMakeFiles/nvim.dir/option.c.o
ninja: build stopped: subcommand failed.
make: *** [nvim] Error 1
```


In the Rosetta iTerm:
```
brew install ninja libtool automake cmake pkg-config gettext
brew install -s --HEAD neovim
```

Which gives this error:
```
cd /tmp/neovim-20201214-72905-1msxk8r/build/src/nvim && /usr/bin/clang -DINCLUDE_GENERATED_DECLARATIONS -DMIN_LOG_LEVEL=3 -DNVIM_MSGPACK_HAS_FLOAT32 -D_GNU_SOURCE -Dnvim_EXPORTS -I/tmp/neovim-20201214-72905-1msxk8r/build/config -I/tmp/neovim-20201214-72905-1msxk8r/src -I/tmp/neovim-20201214-72905-1msxk8r/build/src/nvim/auto -I/tmp/neovim-20201214-72905-1msxk8r/build/include -I/usr/local/include/luajit-2.0 -isystem /usr/local/include -isystem /tmp/neovim-20201214-72905-1msxk8r/deps-build/include -isystem /usr/local/opt/gettext/include -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNDEBUG -flto=thin -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -o CMakeFiles/nvim.dir/os/time.c.o -c /tmp/neovim-20201214-72905-1msxk8r/src/nvim/os/time.c
/tmp/neovim-20201214-72905-1msxk8r/src/nvim/os/lang.c:63:15: error: use of undeclared identifier 'LC_ALL'; did you mean 'P_ALL'?
    setlocale(LC_ALL, "");
              ^~~~~~
              P_ALL
/Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk/usr/include/sys/wait.h:80:2: note: 'P_ALL' declared here
        P_ALL,
        ^
2 errors generated.
make[2]: *** [src/nvim/CMakeFiles/nvim.dir/os/lang.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 50%] Building C object src/nvim/CMakeFiles/nvim.dir/os/tty.c.o
cd /tmp/neovim-20201214-72905-1msxk8r/build/src/nvim && /usr/bin/clang -DINCLUDE_GENERATED_DECLARATIONS -DMIN_LOG_LEVEL=3 -DNVIM_MSGPACK_HAS_FLOAT32 -D_GNU_SOURCE -Dnvim_EXPORTS -I/tmp/neovim-20201214-72905-1msxk8r/build/config -I/tmp/neovim-20201214-72905-1msxk8r/src -I/tmp/neovim-20201214-72905-1msxk8r/build/src/nvim/auto -I/tmp/neovim-20201214-72905-1msxk8r/build/include -I/usr/local/include/luajit-2.0 -isystem /usr/local/include -isystem /tmp/neovim-20201214-72905-1msxk8r/deps-build/include -isystem /usr/local/opt/gettext/include -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNDEBUG -flto=thin -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -o CMakeFiles/nvim.dir/os/tty.c.o -c /tmp/neovim-20201214-72905-1msxk8r/src/nvim/os/tty.c
make[1]: *** [src/nvim/CMakeFiles/nvim.dir/all] Error 2
make: *** [all] Error 2
```

I can see in the error message for the rosetta version it mentions -arch arm64 so I've definitely screwed things up as the Rosetta one shouldn't be using that.

Yes, I have Xcode installed, just realised I'd never opened it so I did for it to install components.
Exactly the same from terminal, here's the output along with me checking for the variable straight after.

```
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/julianf/Code/neovim/build
touch build/.ran-cmake
ninja  -C build
ninja: Entering directory `build'
[237/291] Building C object src/nvim/CMakeFiles/nvim.dir/os/lang.c.o
FAILED: src/nvim/CMakeFiles/nvim.dir/os/lang.c.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -DINCLUDE_GENERATED_DECLARATIONS -DMIN_LOG_LEVEL=1 -DNVIM_MSGPACK_HAS_FLOAT32 -D_GNU_SOURCE -Dnvim_EXPORTS -Iconfig -I../src -Isrc/nvim/auto -Iinclude -I../.deps/usr/include/luajit-2.1 -isystem ../.deps/usr/include -isystem /usr/local/opt/gettext/include -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -g -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -MD -MT src/nvim/CMakeFiles/nvim.dir/os/lang.c.o -MF src/nvim/CMakeFiles/nvim.dir/os/lang.c.o.d -o src/nvim/CMakeFiles/nvim.dir/os/lang.c.o -c ../src/nvim/os/lang.c
../src/nvim/os/lang.c:63:5: error: implicit declaration of function 'setlocale' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    setlocale(LC_ALL, "");
    ^
../src/nvim/os/lang.c:63:15: error: use of undeclared identifier 'LC_ALL'; did you mean 'P_ALL'?
    setlocale(LC_ALL, "");
              ^~~~~~
              P_ALL
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/wait.h:80:2: note: 'P_ALL' declared here
        P_ALL,
        ^
2 errors generated.
[246/291] Building C object src/nvim/CMakeFiles/nvim.dir/option.c.o
ninja: build stopped: subcommand failed.
make: *** [nvim] Error 1
Mac-mini neovim % echo $LC_ALL                        
en_GB.UTF-8
```

I thought earlier I'd managed to get it to show when you run ```locale``` but it isn't now and my history doesn't show me what I did, so ignore that part as no idea if I really did it.

I installed Xcode 12.3 from the App Store.
Manually downloaded Command Line Tools from https://developer.apple.com/download/more/
After installing the command line tools I ran `sudo xcode-select -s /Library/Developer/CommandLineTools`
That was in a normal iTerm zsh running natively.
Cloned neovim with `git clone git@github.com:neovim/neovim`
cd into neovim and ran `CMAKE_BUILD_TYPE=RelWithDebInfo make`

Got the same error:
```
../src/nvim/os/lang.c:63:5: error: implicit declaration of function 'setlocale' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    setlocale(LC_ALL, "");
```

Tried exactly the same on a version of iTerm running under Rosetta, used a simple make and the outcome was exactly the same.

--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -7,6 +7,8 @@
 
 #include <math.h>
 
+#include "auto/config.h"
+
 #ifdef HAVE_LOCALE_H
 # include <locale.h>
 #endif

--- a/src/nvim/os/lang.c
+++ b/src/nvim/os/lang.c
@@ -8,9 +8,12 @@
 # undef Boolean
 #endif
 
+#include "auto/config.h"
+
 #ifdef HAVE_LOCALE_H
 # include <locale.h>
 #endif
+
 #include "nvim/os/lang.h"
 #include "nvim/os/os.h"
 

--- a/src/nvim/os/users.c
+++ b/src/nvim/os/users.c
@@ -5,6 +5,8 @@
 
 #include <uv.h>
 
+#include "auto/config.h"
+
 #include "nvim/ascii.h"
 #include "nvim/os/os.h"
 #include "nvim/garray.h"

--- a/src/nvim/undo.c
+++ b/src/nvim/undo.c
@@ -82,6 +82,8 @@
 #include <string.h>
 #include <fcntl.h>
 
+#include "auto/config.h"
+
 #include "nvim/buffer.h"
 #include "nvim/ascii.h"
 #include "nvim/change.h"

